src/bun.js/webcore/body.zig

```diff
@@ -1258,3 +1263,345 @@ pub fn BodyMixin(comptime Type: type) type {
+        stream.value.ensureStillAlive();
+        var allocator = sink.allocator;
+        var buffer_stream = try allocator.create(ArrayBufferSink.JSSink);
+        var globalThis = sink.global;
+        buffer_stream.* = ArrayBufferSink.JSSink{
+            .sink = ArrayBufferSink{
+                .bytes = bun.ByteList.init(&.{}),
+                .allocator = allocator,
+                .next = null,
+            },
+        };
+        var signal = &buffer_stream.sink.signal;
+        sink.js_sink = buffer_stream;
+
+        signal.* = ArrayBufferSink.JSSink.SinkSignal.init(JSValue.zero);
+
+        // explicitly set it to a dead pointer
+        // we use this memory address to disable signals being sent
+        signal.clear();
+        std.debug.assert(signal.isDead());
+
+        const assignment_result: JSValue = ArrayBufferSink.JSSink.assignToStream(
+            globalThis,
+            stream.value,
+            buffer_stream,
+            @as(**anyopaque, @ptrCast(&signal.ptr)),
+        );
+
+        assignment_result.ensureStillAlive();
+
+        // assert that it was updated
+        std.debug.assert(!signal.isDead());
+
+        if (assignment_result.toError() != null) {
+            return error.PipeFailed;
+        }
+
+        if (!assignment_result.isEmptyOrUndefinedOrNull()) {
+            assignment_result.ensureStillAlive();
+            // it returns a Promise when it goes through ReadableStreamDefaultReader
+            if (assignment_result.asAnyPromise()) |promise| {
+                switch (promise.status(globalThis.vm())) {
+                    .Pending => {
+                        assignment_result.then(
+                            globalThis,
+                            sink,
+                            onResolveStream,
+                            onRejectStream,
+                        );
+                    },
+                    .Fulfilled => {
+                        defer stream.value.unprotect();
+
+                        sink.handleResolveStream(false);
+                    },
+                    .Rejected => {
+                        defer stream.value.unprotect();
+
+                        sink.handleRejectStream(promise.result(globalThis.vm()), false);
+                    },
+                }
+                return;
+            }
+        }
+
+        return error.PipeFailed;
+    }
+
+    fn bufferLockedBodyValue(sink: *@This(), value: *JSC.WebCore.Body.Value) !void {
+        std.debug.assert(value.* == .Locked);
+        const locked = &value.Locked;
+        if (locked.readable) |stream_| {
+            const stream: JSC.WebCore.ReadableStream = stream_;
+            stream.value.ensureStillAlive();
+            // keep the stream alive until we're done with it
+            sink.readable_stream_ref = try JSC.WebCore.ReadableStream.Strong.init(stream, sink.global);
+
+            value.* = .{ .Used = {} };
+
+            if (stream.isLocked(sink.global)) {
```

Comment:
This leaks memory when the stream is already locked. 

sink.readable_stream_ref needs to only be created after used.

Reply:
fixed in https://github.com/oven-sh/bun/pull/4418/commits/7cd66913e79f7c27d13efeb30801dfcbbabe45e6